File: lib/codegen/meta-python/gen_settings.py
"""
Generate sources with settings.
"""
from __future__ import absolute_import
import srcgen
from unique_table import UniqueSeqTable
import constant_hash
from cdsl import camel_case
from cdsl.settings import BoolSetting, NumSetting, EnumSetting
from base import settings

try:
    from typing import Sequence, Set, Tuple, List, Union, TYPE_CHECKING  # noqa
    if TYPE_CHECKING:
        from cdsl.isa import TargetISA  # noqa
        from cdsl.settings import Setting, Preset, SettingGroup  # noqa
        from cdsl.predicates import Predicate, PredContext  # noqa
except ImportError:
    pass


def gen_to_and_from_str(ty, values, fmt):
    # type: (str, Tuple[str, ...], srcgen.Formatter) -> None
    """
    Emit Display and FromStr implementations for enum settings.
    """
    with fmt.indented('impl fmt::Display for {} {{'.format(ty), '}'):
        with fmt.indented(
                'fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {',
                '}'):
            with fmt.indented('f.write_str(match *self {', '})'):
                for v in values:
                    fmt.line('{}::{} => "{}",'
                             .format(ty, camel_case(v), v))

    with fmt.indented('impl str::FromStr for {} {{'.format(ty), '}'):
        fmt.line('type Err = ();')
        with fmt.indented(
                'fn from_str(s: &str) -> Result<Self, Self::Err> {',
                '}'):
            with fmt.indented('match s {', '}'):
                for v in values:
                    fmt.line('"{}" => Ok({}::{}),'
                             .format(v, ty, camel_case(v)))
                fmt.line('_ => Err(()),')


def gen_enum_types(sgrp, fmt):
    # type: (SettingGroup, srcgen.Formatter) -> None
    """
    Emit enum types for any enum settings.
    """
    for setting in sgrp.settings:
        if not isinstance(setting, EnumSetting):
            continue
        ty = camel_case(setting.name)
        fmt.doc_comment('Values for `{}`.'.format(setting))
        fmt.line('#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]')
        with fmt.indented('pub enum {} {{'.format(ty), '}'):
            for v in setting.values:
                fmt.doc_comment('`{}`.'.format(v))
                fmt.line(camel_case(v) + ',')

        gen_to_and_from_str(ty, setting.values, fmt)


def gen_getter(setting, sgrp, fmt):
    # type: (Setting, SettingGroup, srcgen.Formatter) -> None
    """
    Emit a getter function for `setting`.
    """
    fmt.doc_comment(setting.__doc__)

    if isinstance(setting, BoolSetting):
        proto = 'pub fn {}(&self) -> bool'.format(setting.name)
        with fmt.indented(proto + ' {', '}'):
            fmt.line(
                    'self.numbered_predicate({})'
                    .format(sgrp.predicate_number[setting]))
    elif isinstance(setting, NumSetting):
        proto = 'pub fn {}(&self) -> u8'.format(setting.name)
        with fmt.indented(proto + ' {', '}'):
            fmt.line('self.bytes[{}]'.format(setting.byte_offset))
    elif isinstance(setting, EnumSetting):
        ty = camel_case(setting.name)
        proto = 'pub fn {}(&self) -> {}'.format(setting.name, ty)
        with fmt.indented(proto + ' {', '}'):
            m = srcgen.Match('self.bytes[{}]'.format(setting.byte_offset))
            for i, v in enumerate(setting.values):
                m.arm(str(i), [], '{}::{}'.format(ty, camel_case(v)))
            m.arm('_', [], 'panic!("Invalid enum value")')
            fmt.match(m)
    else:
        raise AssertionError("Unknown setting kind")


def gen_pred_getter(name, pred, sgrp, fmt):
    # type: (str, Predicate, SettingGroup, srcgen.Formatter) -> None
    """
    Emit a getter for a named pre-computed predicate.
    """
    fmt.doc_comment('Computed predicate `{}`.'.format(pred.rust_predicate(0)))
    proto = 'pub fn {}(&self) -> bool'.format(name)
    with fmt.indented(proto + ' {', '}'):
        fmt.line(
                'self.numbered_predicate({})'
                .format(sgrp.predicate_number[pred]))


def gen_getters(sgrp, fmt):
    # type: (SettingGroup, srcgen.Formatter) -> None
    """
    Emit getter functions for all the settings in fmt.
    """
    fmt.doc_comment("User-defined settings.")
    fmt.line('#[allow(dead_code)]')
    with fmt.indented('impl Flags {', '}'):
        fmt.doc_comment('Get a view of the boolean predicates.')
        with fmt.indented(
                'pub fn predicate_view(&self) -> ::settings::PredicateView {',
                '}'):
            fmt.format(
                    '::settings::PredicateView::new(&self.bytes[{}..])',
                    sgrp.boolean_offset)
        if sgrp.settings:
            fmt.doc_comment('Dynamic numbered predicate getter.')
            with fmt.indented(
                    'fn numbered_predicate(&self, p: usize) -> bool {', '}'):
                fmt.line(
                        'self.bytes[{} + p / 8] & (1 << (p % 8)) != 0'
                        .format(sgrp.boolean_offset))
        for setting in sgrp.settings:
            gen_getter(setting, sgrp, fmt)
        for name, pred in sgrp.named_predicates.items():
            gen_pred_getter(name, pred, sgrp, fmt)


def gen_descriptors(sgrp, fmt):
    # type: (SettingGroup, srcgen.Formatter) -> None
    """
    Generate the DESCRIPTORS, ENUMERATORS, and PRESETS tables.
    """

    enums = UniqueSeqTable()

    with fmt.indented(
            'static DESCRIPTORS: [detail::Descriptor; {}] = ['
            .format(len(sgrp.settings) + len(sgrp.presets)),
            '];'):
        for idx, setting in enumerate(sgrp.settings):
            setting.descriptor_index = idx
            with fmt.indented('detail::Descriptor {', '},'):
                fmt.line('name: "{}",'.format(setting.name))
                fmt.line('offset: {},'.format(setting.byte_offset))
                if isinstance(setting, BoolSetting):
                    fmt.line(
                            'detail: detail::Detail::Bool {{ bit: {} }},'
                            .format(setting.bit_offset))
                elif isinstance(setting, NumSetting):
                    fmt.line('detail: detail::Detail::Num,')
                elif isinstance(setting, EnumSetting):
                    offs = enums.add(setting.values)
                    fmt.line(
                            'detail: detail::Detail::Enum ' +
                            '{{ last: {}, enumerators: {} }},'
                            .format(len(setting.values)-1, offs))
                else:
                    raise AssertionError("Unknown setting kind")

        for idx, preset in enumerate(sgrp.presets):
            preset.descriptor_index = len(sgrp.settings) + idx
            with fmt.indented('detail::Descriptor {', '},'):
                fmt.line('name: "{}",'.format(preset.name))
                fmt.line('offset: {},'.format(idx * sgrp.settings_size))
                fmt.line('detail: detail::Detail::Preset,')

    with fmt.indented(
            'static ENUMERATORS: [&str; {}] = ['
            .format(len(enums.table)),
            '];'):
        for txt in enums.table:
            fmt.line('"{}",'.format(txt))

    def hash_setting(s):
        # type: (Union[Setting, Preset]) -> int
        return constant_hash.simple_hash(s.name)

    hash_elems = []  # type: List[Union[Setting, Preset]]
    hash_elems.extend(sgrp.settings)
    hash_elems.extend(sgrp.presets)
    hash_table = constant_hash.compute_quadratic(hash_elems, hash_setting)
    with fmt.indented(
            'static HASH_TABLE: [u16; {}] = ['
            .format(len(hash_table)),
            '];'):
        for h in hash_table:
            if h is None:
                fmt.line('0xffff,')
            else:
                fmt.line('{},'.format(h.descriptor_index))

    with fmt.indented(
            'static PRESETS: [(u8, u8); {}] = ['
            .format(len(sgrp.presets) * sgrp.settings_size),
            '];'):
        for preset in sgrp.presets:
            fmt.comment(preset.name)
            for mask, value in preset.layout():
                fmt.format('(0b{:08b}, 0b{:08b}),', mask, value)


def gen_template(sgrp, fmt):
    # type: (SettingGroup, srcgen.Formatter) -> None
    """
    Emit a Template constant.
    """
    v = [0] * sgrp.settings_size
    for setting in sgrp.settings:
        v[setting.byte_offset] |= setting.default_byte()

    with fmt.indented(
            'static TEMPLATE: detail::Template = detail::Template {', '};'):
        fmt.line('name: "{}",'.format(sgrp.name))
        fmt.line('descriptors: &DESCRIPTORS,')
        fmt.line('enumerators: &ENUMERATORS,')
        fmt.line('hash_table: &HASH_TABLE,')
        vs = ', '.join('{:#04x}'.format(x) for x in v)
        fmt.line('defaults: &[{}],'.format(vs))
        fmt.line('presets: &PRESETS,')

    fmt.doc_comment(
            'Create a `settings::Builder` for the {} settings group.'
            .format(sgrp.name))
    with fmt.indented('pub fn builder() -> Builder {', '}'):
        fmt.line('Builder::new(&TEMPLATE)')


def gen_display(sgrp, fmt):
    # type: (SettingGroup, srcgen.Formatter) -> None
    """
    Generate the Display impl for Flags.
    """
    with fmt.indented('impl fmt::Display for Flags {', '}'):
        with fmt.indented(
                'fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {',
                '}'):
            fmt.line('writeln!(f, "[{}]")?;'.format(sgrp.name))
            with fmt.indented('for d in &DESCRIPTORS {', '}'):
                with fmt.indented('if !d.detail.is_preset() {', '}'):
                    fmt.line('write!(f, "{} = ", d.name)?;')
                    fmt.line(
                            'TEMPLATE.format_toml_value(d.detail, ' +
                            'self.bytes[d.offset as usize], f)?;')
                    fmt.line('writeln!(f)?;')
            fmt.line('Ok(())')


def gen_constructor(sgrp, parent, fmt):
    # type: (SettingGroup, PredContext, srcgen.Formatter) -> None
    """
    Generate a Flags constructor.
    """

    with fmt.indented('impl Flags {', '}'):
        args = 'builder: Builder'
        if sgrp.parent:
            p = sgrp.parent
            args = '{}: &{}::Flags, {}'.format(p.name, p.qual_mod, args)
        fmt.doc_comment('Create flags {} settings group.'.format(sgrp.name))
        fmt.line('#[allow(unused_variables)]')
        with fmt.indented(
                'pub fn new({}) -> Self {{'.format(args), '}'):
            fmt.line('let bvec = builder.state_for("{}");'.format(sgrp.name))
            fmt.line(
                'let mut {} = Self {{ bytes: [0; {}] }};'
                .format(sgrp.name, sgrp.byte_size()))
            fmt.line(
                'debug_assert_eq!(bvec.len(), {});'
                .format(sgrp.settings_size))
            fmt.line(
                '{}.bytes[0..{}].copy_from_slice(&bvec);'
                .format(sgrp.name, sgrp.settings_size))

            # Now compute the predicates.
            for pred, number in sgrp.predicate_number.items():
                # Don't compute our own settings.
                if number < sgrp.boolean_settings:
                    continue
                fmt.comment('Precompute #{}.'.format(number))
                with fmt.indented(
                        'if {} {{'.format(pred.rust_predicate(0)),
                        '}'):
                    fmt.line(
                            '{}.bytes[{}] |= 1 << {};'
                            .format(
                                sgrp.name,
                                sgrp.boolean_offset + number // 8,
                                number % 8))

            fmt.line(sgrp.name)


def gen_group(sgrp, fmt):
    # type: (SettingGroup, srcgen.Formatter) -> None
    """
    Generate a Flags struct representing `sgrp`.
    """
    fmt.line('#[derive(Clone)]')
    fmt.doc_comment('Flags group `{}`.'.format(sgrp.name))
    with fmt.indented('pub struct Flags {', '}'):
        fmt.line('bytes: [u8; {}],'.format(sgrp.byte_size()))

    gen_constructor(sgrp, None, fmt)
    gen_enum_types(sgrp, fmt)
    gen_getters(sgrp, fmt)
    gen_descriptors(sgrp, fmt)
    gen_template(sgrp, fmt)
    gen_display(sgrp, fmt)


def generate(isas, out_dir):
    # type: (Sequence[TargetISA], str) -> None
    # Generate shared settings.
    fmt = srcgen.Formatter()
    settings.group.qual_mod = 'settings'
    gen_group(settings.group, fmt)
    fmt.update_file('settings.rs', out_dir)

    # Generate ISA-specific settings.
    for isa in isas:
        isa.settings.qual_mod = 'isa::{}::settings'.format(
                isa.settings.name)
        fmt = srcgen.Formatter()
        gen_group(isa.settings, fmt)
        fmt.update_file('settings-{}.rs'.format(isa.name), out_dir)

File: cranelift-codegen/meta/src/gen_settings.rs
use crate::base;
use crate::cdsl::camel_case;
use crate::cdsl::isa::TargetIsa;
use crate::cdsl::settings::{
    BoolSetting, Predicate, Preset, Setting, SettingGroup, SpecificSetting,
};
use crate::constant_hash::{generate_table, simple_hash};
use crate::error;
use crate::srcgen::{Formatter, Match};
use crate::unique_table::UniqueTable;
use std::collections::HashMap;

enum ParentGroup {
    None,
    Shared,
}

/// Emits the constructor of the Flags structure.
fn gen_constructor(group: &SettingGroup, parent: ParentGroup, fmt: &mut Formatter) {
    let args = match parent {
        ParentGroup::None => "builder: Builder",
        ParentGroup::Shared => "shared: &settings::Flags, builder: Builder",
    };
    fmt.line("impl Flags {");
    fmt.indent(|fmt| {
        fmt.doc_comment(&format!("Create flags {} settings group.", group.name));
        fmt.line("#[allow(unused_variables)]");
        fmt.line(&format!("pub fn new({}) -> Self {{", args));
        fmt.indent(|fmt| {
            fmt.line(&format!(
                "let bvec = builder.state_for(\"{}\");",
                group.name
            ));
            fmt.line(&format!(
                "let mut {} = Self {{ bytes: [0; {}] }};",
                group.name,
                group.byte_size()
            ));
            fmt.line(&format!(
                "debug_assert_eq!(bvec.len(), {});",
                group.settings_size
            ));
            fmt.line(&format!(
                "{}.bytes[0..{}].copy_from_slice(&bvec);",
                group.name, group.settings_size
            ));

            // Now compute the predicates.
            for p in &group.predicates {
                fmt.comment(&format!("Precompute #{}.", p.number));
                fmt.line(&format!("if {} {{", p.render(group)));
                fmt.indent(|fmt| {
                    fmt.line(&format!(
                        "{}.bytes[{}] |= 1 << {};",
                        group.name,
                        group.bool_start_byte_offset + p.number / 8,
                        p.number % 8
                    ));
                });
                fmt.line("}");
            }

            fmt.line(group.name);
        });
        fmt.line("}");
    });
    fmt.line("}");
}

/// Emit Display and FromStr implementations for enum settings.
fn gen_to_and_from_str(name: &str, values: &[&'static str], fmt: &mut Formatter) {
    fmt.line(&format!("impl fmt::Display for {} {{", name));
    fmt.indent(|fmt| {
        fmt.line("fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {");
        fmt.indent(|fmt| {
            fmt.line("f.write_str(match *self {");
            fmt.indent(|fmt| {
                for v in values.iter() {
                    fmt.line(&format!("{}::{} => \"{}\",", name, camel_case(v), v));
                }
            });
            fmt.line("})");
        });
        fmt.line("}");
    });
    fmt.line("}");

    fmt.line(&format!("impl str::FromStr for {} {{", name));
    fmt.indent(|fmt| {
        fmt.line("type Err = ();");
        fmt.line("fn from_str(s: &str) -> Result<Self, Self::Err> {");
        fmt.indent(|fmt| {
            fmt.line("match s {");
            fmt.indent(|fmt| {
                for v in values.iter() {
                    fmt.line(&format!("\"{}\" => Ok({}::{}),", v, name, camel_case(v)));
                }
                fmt.line("_ => Err(()),");
            });
            fmt.line("}");
        });
        fmt.line("}");
    });
    fmt.line("}");
}

/// Emit real enum for the Enum settings.
fn gen_enum_types(group: &SettingGroup, fmt: &mut Formatter) {
    for setting in group.settings.iter() {
        let values = match setting.specific {
            SpecificSetting::Bool(_) | SpecificSetting::Num(_) => continue,
            SpecificSetting::Enum(ref values) => values,
        };
        let name = camel_case(setting.name);

        fmt.doc_comment(&format!("Values for `{}.{}`.", group.name, setting.name));
        fmt.line("#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]");
        fmt.line(&format!("pub enum {} {{", name));
        fmt.indent(|fmt| {
            for v in values.iter() {
                fmt.doc_comment(&format!("`{}`.", v));
                fmt.line(&format!("{},", camel_case(v)));
            }
        });
        fmt.line("}");

        gen_to_and_from_str(&name, values, fmt);
    }
}

/// Emit a getter function for `setting`.
fn gen_getter(setting: &Setting, fmt: &mut Formatter) {
    fmt.doc_comment(setting.comment);
    match setting.specific {
        SpecificSetting::Bool(BoolSetting {
            predicate_number, ..
        }) => {
            fmt.line(&format!("pub fn {}(&self) -> bool {{", setting.name));
            fmt.indent(|fmt| {
                fmt.line(&format!("self.numbered_predicate({})", predicate_number));
            });
            fmt.line("}");
        }
        SpecificSetting::Enum(ref values) => {
            let ty = camel_case(setting.name);
            fmt.line(&format!("pub fn {}(&self) -> {} {{", setting.name, ty));
            fmt.indent(|fmt| {
                let mut m = Match::new(format!("self.bytes[{}]", setting.byte_offset));
                for (i, v) in values.iter().enumerate() {
                    m.arm(
                        format!("{}", i),
                        vec![],
                        format!("{}::{}", ty, camel_case(v)),
                    );
                }
                m.arm("_", vec![], "panic!(\"Invalid enum value\")");
                fmt.add_match(m);
            });
            fmt.line("}");
        }
        SpecificSetting::Num(_) => {
            fmt.line(&format!("pub fn {}(&self) -> u8 {{", setting.name));
            fmt.indent(|fmt| {
                fmt.line(&format!("self.bytes[{}]", setting.byte_offset));
            });
            fmt.line("}");
        }
    }
}

fn gen_pred_getter(predicate: &Predicate, group: &SettingGroup, fmt: &mut Formatter) {
    fmt.doc_comment(&format!(
        "Computed predicate `{}`.",
        predicate.render(group)
    ));
    fmt.line(&format!("pub fn {}(&self) -> bool {{", predicate.name));
    fmt.indent(|fmt| {
        fmt.line(&format!("self.numbered_predicate({})", predicate.number));
    });
    fmt.line("}");
}

/// Emits getters for each setting value.
fn gen_getters(group: &SettingGroup, fmt: &mut Formatter) {
    fmt.doc_comment("User-defined settings.");
    fmt.line("#[allow(dead_code)]");
    fmt.line("impl Flags {");
    fmt.indent(|fmt| {
        fmt.doc_comment("Get a view of the boolean predicates.");
        fmt.line("pub fn predicate_view(&self) -> ::settings::PredicateView {");
        fmt.indent(|fmt| {
            fmt.line(&format!(
                "::settings::PredicateView::new(&self.bytes[{}..])",
                group.bool_start_byte_offset
            ));
        });
        fmt.line("}");

        if group.settings.len() > 0 {
            fmt.doc_comment("Dynamic numbered predicate getter.");
            fmt.line("fn numbered_predicate(&self, p: usize) -> bool {");
            fmt.indent(|fmt| {
                fmt.line(&format!(
                    "self.bytes[{} + p / 8] & (1 << (p % 8)) != 0",
                    group.bool_start_byte_offset
                ));
            });
            fmt.line("}");
        }

        for setting in &group.settings {
            gen_getter(&setting, fmt);
        }
        for predicate in &group.predicates {
            gen_pred_getter(&predicate, &group, fmt);
        }
    });
    fmt.line("}");
}

#[derive(Hash, PartialEq, Eq)]
enum SettingOrPreset<'a> {
    Setting(&'a Setting),
    Preset(&'a Preset),
}

impl<'a> SettingOrPreset<'a> {
    fn name(&self) -> &str {
        match *self {
            SettingOrPreset::Setting(s) => s.name,
            SettingOrPreset::Preset(p) => p.name,
        }
    }
}

/// Emits DESCRIPTORS, ENUMERATORS, HASH_TABLE and PRESETS.
fn gen_descriptors(group: &SettingGroup, fmt: &mut Formatter) {
    let mut enum_table: UniqueTable<&'static str> = UniqueTable::new();

    let mut descriptor_index_map: HashMap<SettingOrPreset, usize> = HashMap::new();

    // Generate descriptors.
    fmt.line(&format!(
        "static DESCRIPTORS: [detail::Descriptor; {}] = [",
        group.settings.len() + group.presets.len()
    ));
    fmt.indent(|fmt| {
        for (idx, setting) in group.settings.iter().enumerate() {
            fmt.line("detail::Descriptor {");
            fmt.indent(|fmt| {
                fmt.line(&format!("name: \"{}\",", setting.name));
                fmt.line(&format!("offset: {},", setting.byte_offset));
                match setting.specific {
                    SpecificSetting::Bool(BoolSetting { bit_offset, .. }) => {
                        fmt.line(&format!(
                            "detail: detail::Detail::Bool {{ bit: {} }},",
                            bit_offset
                        ));
                    }
                    SpecificSetting::Enum(ref values) => {
                        let offset = enum_table.add(values);
                        fmt.line(&format!(
                            "detail: detail::Detail::Enum {{ last: {}, enumerators: {} }},",
                            values.len() - 1,
                            offset
                        ));
                    }
                    SpecificSetting::Num(_) => {
                        fmt.line("detail: detail::Detail::Num,");
                    }
                }

                descriptor_index_map.insert(SettingOrPreset::Setting(setting), idx);
            });
            fmt.line("},");
        }

        for (idx, preset) in group.presets.iter().enumerate() {
            fmt.line("detail::Descriptor {");
            fmt.indent(|fmt| {
                fmt.line(&format!("name: \"{}\",", preset.name));
                fmt.line(&format!("offset: {},", (idx as u8) * group.settings_size));
                fmt.line("detail: detail::Detail::Preset,");
            });
            fmt.line("},");

            descriptor_index_map.insert(SettingOrPreset::Preset(preset), idx);
        }
    });
    fmt.line("];");

    // Generate enumerators.
    fmt.line(&format!(
        "static ENUMERATORS: [&str; {}] = [",
        enum_table.len()
    ));
    fmt.indent(|fmt| {
        for enum_val in enum_table.iter() {
            fmt.line(&format!("\"{}\",", enum_val));
        }
    });
    fmt.line("];");

    // Generate hash table.
    let mut hash_entries: Vec<SettingOrPreset> = Vec::new();
    hash_entries.extend(
        group
            .settings
            .iter()
            .map(|x| SettingOrPreset::Setting(x))
            .collect::<Vec<SettingOrPreset>>(),
    );
    hash_entries.extend(
        group
            .presets
            .iter()
            .map(|x| SettingOrPreset::Preset(x))
            .collect::<Vec<SettingOrPreset>>(),
    );
    let hash_table = generate_table(&hash_entries, |entry| simple_hash(entry.name()));
    fmt.line(&format!(
        "static HASH_TABLE: [u16; {}] = [",
        hash_table.len()
    ));
    fmt.indent(|fmt| {
        for h in &hash_table {
            match *h {
                Some(setting_or_preset) => fmt.line(&format!(
                    "{},",
                    &descriptor_index_map
                        .get(setting_or_preset)
                        .unwrap()
                        .to_string()
                )),
                None => fmt.line("0xffff,"),
            }
        }
    });
    fmt.line("];");

    // Generate presets.
    fmt.line(&format!(
        "static PRESETS: [(u8, u8); {}] = [",
        group.presets.len()
    ));
    fmt.indent(|fmt| {
        for preset in &group.presets {
            fmt.comment(preset.name);
            for (mask, value) in preset.layout(&group) {
                fmt.line(&format!("(0b{:08b}, 0b{:08b}),", mask, value));
            }
        }
    });
    fmt.line("];");
}

fn gen_template(group: &SettingGroup, fmt: &mut Formatter) {
    let mut default_bytes: Vec<u8> = vec![0; group.settings_size as usize];
    for setting in &group.settings {
        *default_bytes.get_mut(setting.byte_offset as usize).unwrap() |= setting.default_byte();
    }

    let default_bytes: Vec<String> = default_bytes
        .iter()
        .map(|x| format!("{:#04x}", x))
        .collect();
    let default_bytes_str = default_bytes.join(", ");

    fmt.line("static TEMPLATE: detail::Template = detail::Template {");
    fmt.indent(|fmt| {
        fmt.line(&format!("name: \"{}\",", group.name));
        fmt.line("descriptors: &DESCRIPTORS,");
        fmt.line("enumerators: &ENUMERATORS,");
        fmt.line("hash_table: &HASH_TABLE,");
        fmt.line(&format!("defaults: &[{}],", default_bytes_str));
        fmt.line("presets: &PRESETS,");
    });
    fmt.line("};");

    fmt.doc_comment(&format!(
        "Create a `settings::Builder` for the {} settings group.",
        group.name
    ));
    fmt.line("pub fn builder() -> Builder {");
    fmt.indent(|fmt| {
        fmt.line("Builder::new(&TEMPLATE)");
    });
    fmt.line("}");
}

fn gen_display(group: &SettingGroup, fmt: &mut Formatter) {
    fmt.line("impl fmt::Display for Flags {");
    fmt.indent(|fmt| {
        fmt.line("fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {");
        fmt.indent(|fmt| {
            fmt.line(&format!("writeln!(f, \"[{}]\")?;", group.name));
            fmt.line("for d in &DESCRIPTORS {");
            fmt.indent(|fmt| {
                fmt.line("if !d.detail.is_preset() {");
                fmt.indent(|fmt| {
                    fmt.line("write!(f, \"{} = \", d.name)?;");
                    fmt.line(
                        "TEMPLATE.format_toml_value(d.detail, self.bytes[d.offset as usize], f)?;",
                    );
                    fmt.line("writeln!(f)?;");
                });
                fmt.line("}");
            });
            fmt.line("}");
            fmt.line("Ok(())");
        });
        fmt.line("}")
    });
    fmt.line("}");
}

fn gen_group(group: &SettingGroup, parent: ParentGroup, fmt: &mut Formatter) {
    // Generate struct.
    fmt.line("#[derive(Clone)]");
    fmt.doc_comment(&format!("Flags group `{}`.", group.name));
    fmt.line("pub struct Flags {");
    fmt.indent(|fmt| {
        fmt.line(&format!("bytes: [u8; {}],", group.byte_size()));
    });
    fmt.line("}");

    gen_constructor(group, parent, fmt);
    gen_enum_types(group, fmt);
    gen_getters(group, fmt);
    gen_descriptors(group, fmt);
    gen_template(group, fmt);
    gen_display(group, fmt);
}

pub fn generate_common(filename: &str, out_dir: &str) -> Result<SettingGroup, error::Error> {
    let settings = base::settings::generate();
    let mut fmt = Formatter::new();
    gen_group(&settings, ParentGroup::None, &mut fmt);
    fmt.update_file(filename, out_dir)?;
    Ok(settings)
}

pub fn generate(isa: &TargetIsa, prefix: &str, out_dir: &str) -> Result<(), error::Error> {
    let mut fmt = Formatter::new();
    gen_group(&isa.settings, ParentGroup::Shared, &mut fmt);
    fmt.update_file(&format!("{}-{}.rs", prefix, isa.name), out_dir)?;
    Ok(())
}
